lib/xenops_server.ml

```diff
@@ -1844,6 +1840,45 @@ module VGPU = struct
 			(fun () ->
 				debug "VGPU.list %s" vm;
 				DB.list vm) ()
+
+	let receive_vgpu uri cookies s context : unit =
+		let module Request = Cohttp.Request.Make(Cohttp_posix_io.Unbuffered_IO) in
+		let module Response = Cohttp.Response.Make(Cohttp_posix_io.Unbuffered_IO) in
+		let dbg = List.assoc "dbg" cookies in
+		Debug.with_thread_associated dbg
+		(fun () ->
+			let is_localhost, vgpu_id = Debug.with_thread_associated dbg
+				debug "VGPU.receive_vgpu";
+				let remote_instance = List.assoc "instance_id" cookies in
+				let is_localhost = instance_id = remote_instance in
+				(* The URI is /service/xenops/migrate-vgpu/id *)
+				let bits = Stdext.Xstringext.String.split '/' (Uri.path uri) in
+				let vgpu_id_str = bits |> List.rev |> List.hd in
+				debug "VGPU.receive_vgpu vgpu_id_str = %s is_localhost = %b" vgpu_id_str is_localhost;
+				is_localhost, VGPU_DB.id_of_string vgpu_id_str
+			in
+			match context.transferred_fd with
+			| Some transferred_fd ->
+				let op = Atomic (VGPU_restore(vgpu_id, FD transferred_fd)) in
```

Comment:
Or instead of wrapping this `atomic` in an `operation` should we keep it as-is and then instead of `{queue|immediate}_operation` call `perform_atomics` or `queue_atomic_int` or something?